File: SimpleMoxieSwitcher-Windows/SimpleMoxieSwitcher/Services/MQTTService.cs
using MQTTnet;
using MQTTnet.Client;
using MQTTnet.Extensions.ManagedClient;
using MQTTnet.Packets;
using MQTTnet.Protocol;
using Newtonsoft.Json;
using SimpleMoxieSwitcher.Services.Interfaces;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;

namespace SimpleMoxieSwitcher.Services;

public class MQTTService : IMQTTService
{
    private readonly IManagedMqttClient _mqttClient;
    private readonly string _deviceId = "d_openmoxie_windows";
    private readonly string _clientId;

    public event EventHandler<MqttMessageEventArgs>? MessageReceived;
    public event EventHandler<bool>? ConnectionStatusChanged;

    public bool IsConnected => _mqttClient.IsConnected;

    public MQTTService()
    {
        _clientId = $"SimpleMoxieSwitcher-{Guid.NewGuid()}";
        _mqttClient = new MqttFactory().CreateManagedMqttClient();
        SetupMQTT();
    }

    private void SetupMQTT()
    {
        _mqttClient.ApplicationMessageReceivedAsync += OnMessageReceivedAsync;
        _mqttClient.ConnectedAsync += OnConnectedAsync;
        _mqttClient.DisconnectedAsync += OnDisconnectedAsync;
    }

    public async Task ConnectAsync()
    {
        var options = new ManagedMqttClientOptionsBuilder()
            .WithAutoReconnectDelay(TimeSpan.FromSeconds(5))
            .WithClientOptions(new MqttClientOptionsBuilder()
                .WithTcpServer("localhost", 8883)
                .WithClientId(_clientId)
                .WithCredentials("unknown", "")
                .WithTls(new MqttClientOptionsBuilderTlsParameters
                {
                    UseTls = true,
                    AllowUntrustedCertificates = true,
                    IgnoreCertificateChainErrors = true,
                    IgnoreCertificateRevocationErrors = true
                })
                .WithKeepAlivePeriod(TimeSpan.FromSeconds(60))
                .Build())
            .Build();

        await _mqttClient.StartAsync(options);
    }

    public async Task DisconnectAsync()
    {
        await _mqttClient.StopAsync();
    }

    public async Task SendCommandAsync(string command, string speech)
    {
        var payload = new Dictionary<string, object>
        {
            ["event_id"] = Guid.NewGuid().ToString(),
            ["command"] = string.IsNullOrEmpty(speech) ? "prompt" : "continue",
            ["speech"] = speech,
            ["backend"] = "router",
            ["module_id"] = "OPENMOXIE_CHAT",
            ["content_id"] = "default"
        };

        var jsonPayload = JsonConvert.SerializeObject(payload);
        var topic = $"/devices/{_deviceId}/events/remote-chat";

        var message = new MqttApplicationMessageBuilder()
            .WithTopic(topic)
            .WithPayload(jsonPayload)
            .WithQualityOfServiceLevel(MqttQualityOfServiceLevel.AtLeastOnce)
            .Build();

        await _mqttClient.EnqueueAsync(message);
        Console.WriteLine($"📤 Sent remote-chat event: {speech}");
    }

    public async Task PublishAsync(string topic, string message)
    {
        var mqttMessage = new MqttApplicationMessageBuilder()
            .WithTopic(topic)
            .WithPayload(message)
            .WithQualityOfServiceLevel(MqttQualityOfServiceLevel.AtLeastOnce)
            .Build();

        await _mqttClient.EnqueueAsync(mqttMessage);
    }

    private async Task OnConnectedAsync(MqttClientConnectedEventArgs e)
    {
        Console.WriteLine("✅ MQTT connected successfully");

        // Subscribe to command responses from OpenMoxie
        var topics = new List<MqttTopicFilter>
        {
            new MqttTopicFilterBuilder()
                .WithTopic($"/devices/{_deviceId}/commands/remote_chat")
                .WithQualityOfServiceLevel(MqttQualityOfServiceLevel.AtLeastOnce)
                .Build(),
            new MqttTopicFilterBuilder()
                .WithTopic($"/devices/{_deviceId}/commands/+")
                .WithQualityOfServiceLevel(MqttQualityOfServiceLevel.AtLeastOnce)
                .Build(),
            new MqttTopicFilterBuilder()
                .WithTopic($"/devices/{_deviceId}/wakeword")
                .WithQualityOfServiceLevel(MqttQualityOfServiceLevel.AtLeastOnce)
                .Build()
        };

        await _mqttClient.SubscribeAsync(topics);
        Console.WriteLine("📡 Subscribed to command topics");

        ConnectionStatusChanged?.Invoke(this, true);
    }

    private Task OnDisconnectedAsync(MqttClientDisconnectedEventArgs e)
    {
        if (e.Exception != null)
        {
            Console.WriteLine($"MQTT disconnected with error: {e.Exception.Message}");
        }
        else
        {
            Console.WriteLine("MQTT disconnected normally");
        }

        ConnectionStatusChanged?.Invoke(this, false);
        return Task.CompletedTask;
    }

    private Task OnMessageReceivedAsync(MqttApplicationMessageReceivedEventArgs e)
    {
        var topic = e.ApplicationMessage.Topic;
        var payload = Encoding.UTF8.GetString(e.ApplicationMessage.Payload ?? Array.Empty<byte>());

        Console.WriteLine($"📥 MQTT received on {topic}");

        // Parse OpenMoxie remote_chat responses
        if (topic.Contains("/commands/remote_chat"))
        {
            try
            {
                dynamic? json = JsonConvert.DeserializeObject(payload);
                if (json != null)
                {
                    // Extract the response text
                    if (json.output?.text != null)
                    {
                        Console.WriteLine($"🤖 Moxie response: {json.output.text}");
                    }

                    // Extract response actions (for face control, movements, etc.)
                    if (json.response_actions != null)
                    {
                        foreach (var action in json.response_actions)
                        {
                            if (action.action != null)
                            {
                                Console.WriteLine($"🎬 Action: {action.action}");

                                // Handle execute actions (face control, movements)
                                if (action.action == "execute" && action.function_id != null)
                                {
                                    Console.WriteLine($"⚙️ Execute function: {action.function_id}");
                                    if (action.function_args != null)
                                    {
                                        Console.WriteLine($"   Args: {action.function_args}");
                                    }
                                }
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error parsing MQTT message: {ex.Message}");
            }
        }

        Console.WriteLine($"   Raw: {payload}");

        MessageReceived?.Invoke(this, new MqttMessageEventArgs
        {
            Topic = topic,
            Payload = payload
        });

        return Task.CompletedTask;
    }
}

public class MqttMessageEventArgs : EventArgs
{
    public string Topic { get; set; } = string.Empty;
    public string Payload { get; set; } = string.Empty;
}

File: SimpleMoxieSwitcher/Sources/SimpleMoxieSwitcher/Services/MQTTService.swift
import Foundation
import CocoaMQTT

final class MQTTService: MQTTServiceProtocol {
    private var mqtt: CocoaMQTT?
    private let deviceId = "d_openmoxie_ios"
    private let clientId: String

    init() {
        self.clientId = "SimpleMoxieSwitcher-\(UUID().uuidString)"
        setupMQTT()
    }

    private func setupMQTT() {
        let host = AppConfig.mqttHost
        let port = UInt16(AppConfig.mqttPort)
        let useTLS = AppConfig.mqttUseTLS

        mqtt = CocoaMQTT(clientID: clientId, host: host, port: port)
        guard let mqtt = mqtt else { return }

        mqtt.username = "unknown"
        mqtt.password = ""
        mqtt.keepAlive = 60
        mqtt.autoReconnect = true
        mqtt.enableSSL = useTLS
        mqtt.allowUntrustCACertificate = useTLS  // Allow self-signed certificates when TLS is on

        // Set delegate to handle connection events
        mqtt.delegate = self

        print("MQTT configured for host: \(host):\(port) TLS=\(useTLS)")
    }

    func connect() {
        _ = mqtt?.connect()
    }

    func disconnect() {
        mqtt?.disconnect()
    }

    func sendCommand(_ command: String, speech: String) {
        guard let mqtt = mqtt else { return }

        // OpenMoxie expects remote-chat events, not volley commands
        // The proper format is:
        // - Send TO: /devices/{deviceId}/events/remote-chat
        // - Receive FROM: /devices/{deviceId}/commands/remote_chat

        let payload: [String: Any] = [
            "event_id": UUID().uuidString,
            "command": speech.isEmpty ? "prompt" : "continue",
            "speech": speech,
            "backend": "router",
            "module_id": "OPENMOXIE_CHAT",
            "content_id": "default"
        ]

        if let jsonData = try? JSONSerialization.data(withJSONObject: payload),
           let jsonString = String(data: jsonData, encoding: .utf8) {
            // Send to events/remote-chat topic (this is what OpenMoxie expects)
            let topic = "/devices/\(deviceId)/events/remote-chat"
            mqtt.publish(topic, withString: jsonString, qos: .qos1)
            print("📤 Sent remote-chat event: \(speech)")
        }
    }

    func publish(topic: String, message: String) {
        mqtt?.publish(topic, withString: message, qos: .qos1)
    }
}

// MARK: - CocoaMQTTDelegate
extension MQTTService: CocoaMQTTDelegate {
    func mqtt(_ mqtt: CocoaMQTT, didConnectAck ack: CocoaMQTTConnAck) {
        if ack == .accept {
            print("✅ MQTT connected successfully")
            // Subscribe to command responses from OpenMoxie
            mqtt.subscribe("/devices/\(deviceId)/commands/remote_chat", qos: .qos1)
            mqtt.subscribe("/devices/\(deviceId)/commands/+", qos: .qos1)
            // Subscribe to wake word topic
            mqtt.subscribe("/devices/\(deviceId)/wakeword", qos: .qos1)
            print("📡 Subscribed to command topics")
        } else {
            print("⚠️ MQTT connection failed: \(ack)")
        }
    }

    func mqtt(_ mqtt: CocoaMQTT, didStateChangeTo state: CocoaMQTTConnState) {
        print("MQTT state changed to: \(state)")
    }

    func mqtt(_ mqtt: CocoaMQTT, didPublishMessage message: CocoaMQTTMessage, id: UInt16) {
        print("MQTT published message with id: \(id)")
    }

    func mqtt(_ mqtt: CocoaMQTT, didPublishAck id: UInt16) {
        // Message acknowledged
    }

    func mqtt(_ mqtt: CocoaMQTT, didReceiveMessage message: CocoaMQTTMessage, id: UInt16) {
        if let messageString = message.string {
            print("📥 MQTT received on \(message.topic)")

            // Parse OpenMoxie remote_chat responses
            if message.topic.contains("/commands/remote_chat") {
                if let data = messageString.data(using: .utf8),
                   let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {

                    // Extract the response text
                    if let output = json["output"] as? [String: Any],
                       let text = output["text"] as? String {
                        print("🤖 Moxie response: \(text)")
                    }

                    // Extract response actions (for face control, movements, etc.)
                    if let responseActions = json["response_actions"] as? [[String: Any]] {
                        for action in responseActions {
                            if let actionName = action["action"] as? String {
                                print("🎬 Action: \(actionName)")

                                // Handle execute actions (face control, movements)
                                if actionName == "execute",
                                   let functionId = action["function_id"] as? String {
                                    print("⚙️ Execute function: \(functionId)")
                                    if let functionArgs = action["function_args"] as? [String: Any] {
                                        print("   Args: \(functionArgs)")
                                    }
                                }
                            }
                        }
                    }
                }
            }

            print("   Raw: \(messageString)")
        }
    }

    func mqtt(_ mqtt: CocoaMQTT, didSubscribeTopics success: NSDictionary, failed: [String]) {
        print("MQTT subscribed to topics: \(success)")
        if !failed.isEmpty {
            print("MQTT failed to subscribe to: \(failed)")
        }
    }

    func mqtt(_ mqtt: CocoaMQTT, didUnsubscribeTopics topics: [String]) {
        print("MQTT unsubscribed from topics: \(topics)")
    }

    func mqttDidPing(_ mqtt: CocoaMQTT) {
        // Ping sent
    }

    func mqttDidReceivePong(_ mqtt: CocoaMQTT) {
        // Pong received
    }

    func mqttDidDisconnect(_ mqtt: CocoaMQTT, withError err: Error?) {
        if let error = err {
            print("MQTT disconnected with error: \(error)")
        } else {
            print("MQTT disconnected normally")
        }
    }
}
